src/config/backends/splash-image.ts

```diff
@@ -0,0 +1,196 @@
+	private isPng(buffer: Buffer) {
+		if (!buffer || buffer.length < 8) {
+			return false;
+		}
+
+		return (
+			buffer[0] === 0x89 &&
+			buffer[1] === 0x50 &&
+			buffer[2] === 0x4e &&
+			buffer[3] === 0x47 &&
+			buffer[4] === 0x0d &&
+			buffer[5] === 0x0a &&
+			buffer[6] === 0x1a &&
+			buffer[7] === 0x0a
+		);
+	}
+
+	// Get the file path for the splash image for the underlying
+	// system
+	private async getSplashPath(): Promise<string> {
+		const [
+			// If no logo is defined, assume the file is `balena-logo.png`
+			splashFile = path.join(SplashImage.BASEPATH, 'balena-logo.png'),
+			// Read the splash dir (will throw if the path does not exist)
+		] = await Bluebird.resolve(fs.readdir(SplashImage.BASEPATH))
+			// Check for files in the valid splash filename list
+			.filter((filename) => SplashImage.FILENAMES.includes(filename))
+			// Convert to full path
+			.map((filename) => path.join(SplashImage.BASEPATH, filename));
+
+		return splashFile;
+	}
+
+	// Returns the base64 contents of the splash image
+	private async readSplashImage(where?: string): Promise<string> {
+		// Read from defaultPath unless where is defined
+		where = where ?? (await this.getSplashPath());
+
+		// read the image file...
+		return (await fs.readFile(where)).toString('base64');
+	}
+
+	// Write a splash image provided as a base64 string
+	private async writeSplashImage(image: string, where?: string): Promise<void> {
+		// Read from defaultPath unless where is defined
+		where = where ?? (await this.getSplashPath());
+
+		const buffer = Buffer.from(image, 'base64');
+		if (this.isPng(buffer)) {
+			// Write the buffer to the given location
+			await remountAndWriteAtomic(where, buffer);
+		} else {
+			throw new Error('Splash image should be a base64 encoded PNG image');
+		}
+	}
+
+	private stripPrefix(name: string): string {
+		if (!name.startsWith(SplashImage.PREFIX)) {
+			return name;
+		}
+		return name.substr(SplashImage.PREFIX.length);
+	}
+
+	public createConfigVarName(name: string): string {
+		return `${SplashImage.PREFIX}${name}`;
+	}
+
+	public async getBootConfig(): Promise<ConfigOptions> {
+		try {
+			const defaultImg = await this.readSplashImage(SplashImage.DEFAULT);
+			const img = await this.readSplashImage();
+
+			// If the image is the same as the default image
+			// return nothing
+			if (img !== defaultImg) {
+				return {
+					image: `data:image/png;base64,${img}`,
+				};
+			}
+		} catch (e) {
+			log.warn('Failed to read splash image:', e);
+		}
+		return {};
+	}
+
+	public async initialise(): Promise<SplashImage> {
+		try {
+			await super.initialise();
+
+			// The default boot image file has already
+			// been created
+			if (await fs.exists(SplashImage.DEFAULT)) {
+				return this;
+			}
+
+			// read the existing image file...
+			const image = await this.readSplashImage();
+
+			// write the image to the DEFAULT path
+			await this.writeSplashImage(image, SplashImage.DEFAULT);
```

Comment:
Does this imply that there is always an image on the OS ? I am understanding this as..

1. run readSplashImage 
 a. No where passed so get the where value from getSplashPath function
 b. getSplashPath to me just returns balena-logo.png unless resin-logo.png exists so let's say it returns balena-logo.png
 c. balena-logo.png and now written to balena-logo-default.png

Reply:
Yes, there should always be a resin-logo.png or balena-logo.png on that directory. If there is not, the initialization will fail. On initialization, the backend makes a copy so it can remember the initial value when the the variable is unset. I actually haven't tested that case on a device, it might be checking that no weird loop occurs.

Reply:
Yeah, give that a test. Remember if the OS changes something beneath us we want to be able to easily observe what is causing the weird behavior. The logging in the backend has done a good job of that so far though.

Reply:
This is the output of the supervisor in the case the directory or none of the files exist. I think it's acceptable

```
[debug]   Finished applying target state
[success] Device state apply success
[info]    Applying target state
[warn]    Could not initialise splash image backend Error: ENOENT: no such file or directory, open '/mnt/root/mnt/boot/splash/balena-logo.png'
[warn]    Failed to read splash image: Error: ENOENT: no such file or directory, open '/mnt/root/mnt/boot/splash/balena-logo-default.png'
[warn]    Could not initialise splash image backend Error: ENOENT: no such file or directory, open '/mnt/root/mnt/boot/splash/balena-logo.png'
[debug]   Finished applying target state
[success] Device state apply success
```

Reply:
Right, I remember stepping through the code and reading that exact situation actually. That is enough information for someone to know the next step is to figure out what that file is and why it's missing.